transform/tdrop/tdrop.go

```diff
@@ -39,16 +53,33 @@ func (cfg *Config) VerifyConfig(schema base.LogSchema) error {
 	if err := cfg.Match.VerifyConfig(schema); err != nil {
 		return fmt.Errorf(".match: %w", err)
 	}
+	if cfg.Percentage < 1 || cfg.Percentage > 100 {
+		return fmt.Errorf(".percentage must be between 1 and 100: %d", cfg.Percentage)
+	}
 	if len(cfg.MetricLabel) == 0 {
 		return fmt.Errorf(".metricLabel is unspecified")
 	}
 	return nil
 }
 
 func (tf *dropTransform) Transform(record *base.LogRecord) base.FilterResult {
-	if tf.matcher.Match(record) {
-		tf.counter(record.RawLength)
+	if !tf.matcher.Match(record) {
+		return base.PASS
+	}
+
+	if tf.targetRate == 100 {
+		tf.countDropped(record.RawLength)
 		return base.DROP
 	}
+
+	if tf.totalMatched > 0 && 100*tf.totalDropped/tf.totalMatched < tf.targetRate {
```

Comment:
Would it be more efficient to use a [random number](https://pkg.go.dev/math/rand#Int31n), like

```
rand.Int31n(101) > targetRate    // random number in range 0-100
```

which log rows are retained does not need to be deterministic, we are just doing downsampling here.

Reply:
Verified in `tdropbench_test.go`, benchmarks show 1 (no-rand) : 2 (rand) ns/op. I'll still keep the old code since it's faster and more predictable.

The default rand implementation is confirmed lock free with no external dependencies, so we could also use it in new code.